crates/turbo-tasks-memory/src/scope.rs

```diff
@@ -194,115 +201,101 @@ impl TaskScope {
-                .lock()
-                .children
-                .iter()
-                .copied()
-                .inspect(|i| {
-                    checked_scopes.insert(*i);
-                })
-                .collect::<Vec<_>>();
-            log_scope_update!("has_unfinished_tasks {} -> {:?}", *self.id, queue);
-            while let Some(id) = queue.pop() {
-                match backend.with_scope(id, |scope| {
-                    if scope.unfinished_tasks.load(Ordering::Acquire) != 0 {
-                        let listener = scope.event.listen();
-                        if scope.unfinished_tasks.load(Ordering::Relaxed) != 0 {
-                            return Ok(Some(listener));
-                        }
-                    }
-                    if scope.last_task_finish_generation.load(Ordering::Relaxed) > start_generation
-                    {
-                        return Err(());
-                    }
-                    let scope = scope.state.lock();
-                    queue.extend(
-                        scope
-                            .children
-                            .iter()
-                            .copied()
-                            .filter(|i| checked_scopes.insert(*i)),
-                    );
-                    log_scope_update!(
-                        "has_unfinished_tasks {} -> {:?}",
-                        *scope.id,
-                        scope
-                            .children
-                            .iter()
-                            .copied()
-                            .filter(|i| !checked_scopes.contains(i))
-                            .collect::<Vec<_>>()
-                    );
-                    Ok(None)
-                }) {
-                    Ok(Some(listener)) => {
-                        return Some(listener);
-                    }
-                    Ok(None) => {}
-                    Err(()) => continue 'restart,
+            });
+        }
+    }
+
+    pub fn remove_parent(&self, parent: TaskScopeId, backend: &MemoryBackend) {
+        let remove_unfinished_parent = {
+            let mut state = self.state.lock();
+            state.parents.remove(parent) && state.has_unfinished_tasks
+        };
+        if remove_unfinished_parent {
+            // As we removed a parent while having unfinished tasks we need to decrement the
+            // unfinished task count and potentially update the state
+            backend.with_scope(parent, |parent| {
+                let update = parent.decrement_unfinished_tasks_internal();
+
+                if update {
+                    parent.update_unfinished_state(backend);
                 }
+            });
+        }
+    }
+
+    fn update_unfinished_state(&self, backend: &MemoryBackend) {
+        let mut state = self.state.lock();
+        // we need to load the atomic under the lock to ensure consistency
+        let count = self.unfinished_tasks.load(Ordering::Acquire);
```

Comment:
BRB rewatching [Jon's video on atomics](https://www.youtube.com/watch?v=rMGWeSjctlY) :D

Reply:
Please do, I have no idea if all this is correct...